program/src/main.rs

```diff
@@ -2,104 +2,52 @@
-use alloy::sol;
 use alloy::sol_types::SolType;
 use primitives::get_header_update_verdict;
-use primitives::types::ProofInputs;
-use primitives::types::ProofOutputs;
+use primitives::types::{BridgeCommitmentLeaf, ProofInputs, ProofOutputs};
 use sha2::Sha256;
-use std::collections::HashSet;
-use std::ops::Add;
 use tendermint::{block::Header, merkle::simple_hash_from_byte_vectors};
-use tendermint_light_client_verifier::types::LightBlock;
 use tendermint_light_client_verifier::Verdict;
-type DataRootTuple = sol! {
-    tuple(uint64, bytes32)
-};
 
-/// Compute the data commitment for the supplied headers. Each leaf in the Tendermint Merkle tree
-/// is the SHA256 hash of the concatenation of the block height and the header's last `Commit`.
-/// Excludes the last header's data hash from the commitment to avoid overlapping headers between
-/// commits.
-fn compute_data_commitment(headers: &[Header]) -> [u8; 32] {
-    let mut encoded_data_root_tuples: Vec<Vec<u8>> = Vec::new();
+/// Compute the bridge commitment for the supplied headers. Each leaf in the Tendermint Merkle tree
+/// is the SHA256 hash of the concatenation of the block height and the header's last
+/// `LastResultsHash`. Excludes the last header's last results hash from the commitment to avoid
+/// overlapping headers between commits.
+fn compute_bridge_commitment(headers: &[Header]) -> [u8; 32] {
+    let mut encoded_leaves: Vec<Vec<u8>> = Vec::new();
     // Loop over all headers except the last one.
     for i in 0..headers.len() - 1 {
         let curr_header = &headers[i];
         let next_header = &headers[i + 1];
 
-        // Verify the chain of headers is connected.
-        if curr_header.hash() != next_header.last_block_id.unwrap().hash {
+        // Verify the chain of headers is connected. Skip if genesis block.
+        if curr_header.height.value() > 1
+            && curr_header.hash() != next_header.last_block_id.unwrap().hash
+        {
             panic!("invalid header");
         }
 
-        let data_hash: [u8; 32] = curr_header
-            .data_hash
-            .expect("Header has no data hash.")
+        let last_results_hash: [u8; 32] = curr_header
+            .last_results_hash
+            .ok_or_else(|| {
+                panic!(
+                    "Header at height {} has no last_results_hash",
+                    curr_header.height
+                )
+            })
+            .unwrap()
             .as_bytes()
             .try_into()
             .unwrap();
 
-        // ABI-encode the leaf corresponding to this header, which is a DataRootTuple.
-        let data_root_tuple = DataRootTuple::abi_encode(&(curr_header.height.value(), data_hash));
-        encoded_data_root_tuples.push(data_root_tuple);
+        // ABI-encode the leaf corresponding to this header, which is a BridgeCommitmentLeaf.
+        let encoded_leaf =
+            BridgeCommitmentLeaf::abi_encode(&(&curr_header.height.value(), last_results_hash));
```

Comment:
Just a small diff I noticed. Probably irrelevant.

Original code uses `curr_header.height.value()` but you use `&curr_header.height.value()` (with `&`)

Reply:
it's just passing a reference instead of a value, we are constructing the same merkle root so no difference